### PR TITLE
Namespace sealing POC

### PR DESCRIPTION
**Pull request by wslabosz-reply**
POC for #1357 / #1170 

Currently implemented features:
- per namespace barrier, root key, keyring (data is encrypted only once)
- create sealable namespace (incl. support in cli command) using a seal config
    - currently only supports shamir
    - namespace is unsealed after creation for now, we could reseal after namespace initialization
- store seal config permanently, encrypted with closest parent barrier
- `/sys/raw` works for keys encrypted with non-root barriers

Missing in this POC (not exhaustive):
- loading/unloading of sealed namespaces from storage
- sealing/unsealing lifecycle functionality, api, cli command
- remembering seal shares when calling unseal api
- most other api actions (rekeying, recovery, generate-root, key rotation, seal status, ...)
- most should be straight forward to implement since we are reusing the existing AESGCMBarrier implementation
- handling requests towards sealed namespaces

Example for creating a sealable namespace:

```sh
bao namespace create -seals=seals.json ns1
```

with seals.json (hcl not supported yet):

```json
[
  {
    "type": "shamir",
    "secret_shares": 3,
    "secret_threshold": 2
  }
]
```

/cc @cipherboy, @driif, @voigt, @wslabosz-reply, @rencooo, @satoqz

### CODE REVIEW

File: vault/seal_manager.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+// SealManager is used to provide storage for the seals.
+// It's a singleton that associates seals (configs) to the namespaces.
+type SealManager struct {
+	core    *Core
+	storage logical.Storage
+
+	// lock        sync.RWMutex
+	// invalidated atomic.Bool
+
+	sealsByNamespace map[string][]*Seal
```
Comment (satoqz) [inline vault/seal_manager.go]: `Seal` is an interface that's internally already a pointer.

File: vault/barrier_aes_gcm.go [hunk 2]
```diff
@@ -69,6 +69,8 @@ type AESGCMBarrier struct {
 	l      sync.RWMutex
 	sealed bool
 
+	metaPrefix string
```
Comment (cipherboy) [inline vault/barrier_aes_gcm.go]: I forget, why'd we go this route rather than a proper View?

Comment (cipherboy) [reply vault/barrier_aes_gcm.go]: I think that'd help some of the churn in `vault/barrier_aes_gcm_test.go` tbh, since a `physical.BarrierView` will be implemented by `physical.Barrier`. My 2c.

File: command/namespace_create.go [hunk 3]
```diff
@@ -59,6 +64,14 @@ func (c *NamespaceCreateCommand) Flags() *FlagSets {
 			"This can be specified multiple times to add multiple pieces of metadata.",
 	})
 
+	f.StringVar(&StringVar{
+		Name:       "seals",
+		Target:     &c.flagSealsConfigPath,
+		Completion: complete.PredictFiles("*.json"),
+		Usage: "Path to a JSON file with at least one or several SealConfigs." +
+			"MUST be an JSON array.",
```
Comment (satoqz) [inline command/namespace_create.go]:
```suggestion
		Usage: "Path to a JSON file with a list of namespace seal configurations. " +
			"Must be a JSON array with at least one configuration.",
```

Above `custom-metadata` flag is also missing a space after the first sentence.

File: command/namespace_create.go [hunk 4]
```diff
@@ -96,10 +109,20 @@ func (c *NamespaceCreateCommand) Run(args []string) int {
 		return 2
 	}
 
+	seals, err := c.parseSeals()
+	if err != nil {
+		c.UI.Error(fmt.Sprintf("Error while parsing seal configs: %s", err))
```
Comment (satoqz) [inline command/namespace_create.go]:
```suggestion
		c.UI.Error(fmt.Sprintf("Error parsing seal configs: %s", err))
```

File: command/namespace_create.go [hunk 5]
```diff
@@ -111,5 +134,59 @@ func (c *NamespaceCreateCommand) Run(args []string) int {
...
+	path := c.flagSealsConfigPath
+	if path == "" {
+		return nil, nil
+	}
+	var rawSeals []map[string]interface{}
+
+	absolutePath, err := filepath.Abs(path)
+	if err != nil {
+		return nil, err
+	}
```
Comment (satoqz) [inline command/namespace_create.go]: Is that needed?

File: command/namespace_create.go [hunk 6]
```diff
@@ -111,5 +134,59 @@ func (c *NamespaceCreateCommand) Run(args []string) int {
...
+		// TODO Handle Output for multiple seals
+		keySharesMap := secret.Data["key_shares"].(map[string]interface{})
+		if keySharesMap != nil {
+			defaultKeyShares := keySharesMap["default"].([]interface{})
+			for i, key := range defaultKeyShares {
+				c.UI.Output(fmt.Sprintf("Unseal Key %d: %s", i+1, key))
+			}
+		}
+		secretShares := int(seals[0]["secret_shares"].(float64))
+		secretThreshold := int(seals[0]["secret_threshold"].(float64))
```
Comment (satoqz) [inline command/namespace_create.go]: You're trusting the server to validate that all of these fields do indeed exist. For example, `int(seals[0]["secret_shares"].(float64))` will panic if `"secret_shares"` is not in the map. You can import `vault.SealConfig` here and share parsing/validation logic and only send the seals to the server if they're ok locally. Probably also spares you the `float64` roundtrip.

File: vault/logical_system.go [hunk 7]
```diff
@@ -5893,6 +5893,12 @@ This path responds to the following HTTP methods.
 
 	DELETE /<path>
 		Delete a namespace.
+
+	POST /<name>/seal
+		Seal a namespace.
+
+	POST /<name/unseal
+		Unseal a namespace.
```
Comment (satoqz) [inline vault/logical_system.go]:
```suggestion
	POST /<name>/seal
		Seal a namespace.

	POST /<name>/unseal
		Unseal a namespace.
```

We've been using `<path>` in the namespace backend (also for patterns). I see the appeal in using `<name>` where you can't specify a full namespace path but just an immediate child, but it would be nice to backport this to existing endpoints as well.

File: vault/logical_system_namespaces.go [hunk 8]
```diff
@@ -238,14 +309,19 @@ func (b *SystemBackend) namespacePaths() []*framework.Path {
...
+	}
 	return map[string]any{
 		"uuid":            ns.UUID,
 		"path":            ns.Path,
 		"id":              ns.ID,
 		"tainted":         ns.Tainted,
 		"locked":          ns.Locked,
 		"custom_metadata": ns.CustomMetadata,
+		"key_shares":      keySharesPerSeal,
 	}
```
Comment (satoqz) [inline vault/logical_system_namespaces.go]:
```suggestion
	ret := map[string]any{
		"uuid":            ns.UUID,
		"path":            ns.Path,
		"id":              ns.ID,
		"tainted":         ns.Tainted,
		"locked":          ns.Locked,
		"custom_metadata": ns.CustomMetadata,
	}
	if len(keySharesMap) != 0 {
		ret["key_shares"] = keySharesMap[0]
	}
	return ret
```

File: vault/logical_system_namespaces.go [hunk 9]
```diff
@@ -339,15 +415,50 @@ func (b *SystemBackend) handleNamespacesSet() framework.OperationFunc {
...
+		if new {
+			// TODO(wslabosz): write all the provided configs
+			if len(sealConfigs) > 0 {
+				if err := b.Core.sealManager.SetSeal(ctx, sealConfigs[0], entry); err != nil {
+					return handleError(err)
+				}
+
+				nsSealKeyShares, err := b.Core.sealManager.InitializeBarrier(ctx, entry)
+				if err != nil {
+					return logical.ErrorResponse(fmt.Sprintf("%s", err.Error())), err
```
Comment (satoqz) [inline vault/logical_system_namespaces.go]:
```suggestion
					return handleError(err)
```

File: vault/logical_system_namespaces.go [hunk 10]
```diff
@@ -339,15 +415,50 @@ func (b *SystemBackend) handleNamespacesSet() framework.OperationFunc {
...
+
+		entry, new, err := b.Core.namespaceStore.ModifyNamespaceByPath(ctx, path, func(ctx context.Context, ns *namespace.Namespace) (*namespace.Namespace, error) {
 			ns.CustomMetadata = metadata
 			return ns, nil
 		})
 		if err != nil {
 			return handleError(err)
 		}
+		keySharesMap := make(map[string][]string)
+		if new {
```
Comment (satoqz) [inline vault/logical_system_namespaces.go]: Nit: Could rearrange to early return

File: vault/namespace_store.go [hunk 11]
```diff
@@ -226,66 +226,60 @@ func (ns *NamespaceStore) SetNamespace(ctx context.Context, namespace *namespace
...
-	return ns.setNamespaceLocked(ctx, namespace)
+	// setNamespaceLocked handles unlocking
+	new, err := ns.setNamespaceLocked(ctx, namespace)
+	if err != nil {
+		return err
+	}
+
+	// TODO: might want to support calling (*SealManager).SetSeal here
+	if new {
+		ns.initializeNamespace(ctx, namespace)
```
Comment (satoqz) [inline vault/namespace_store.go]: Unhandled error

File: vault/namespace_store.go [hunk 12]
```diff
@@ -557,20 +545,21 @@ func (ns *NamespaceStore) ModifyNamespaceByPath(ctx context.Context, path string
...
 		}
 
 		// ModifyNamespaceByPath can never modify lock status.
 		entry.UnlockKey = unlockKey
 	}
 
-	// setNamespaceLocked will unlock ns.lock
-	if err := ns.setNamespaceLocked(ctx, entry); err != nil {
-		return nil, err
+	// setNamespaceLocked handles unlocking
```
Comment (satoqz) [inline vault/namespace_store.go]: It really doesn't need to anymore though. Could rearrange callers to do the unlocking now.

File: vault/seal_manager.go [hunk 13]
```diff
@@ -0,0 +1,360 @@
...
+}
+
+func (sm *SealManager) StorageAccessForPath(path string) StorageAccess {
+	_, v, _ := sm.barrierByStoragePath.LongestPrefix(path)
+	if v == nil {
+		return &directStorageAccess{physical: sm.core.physical}
+	}
+	barrier := v.(SecurityBarrier)
+	return &secureStorageAccess{barrier: barrier}
+}
```
Comment (satoqz) [inline vault/seal_manager.go:116]: Would be nice to have a comment somewhere (doesn't have to be on this function) that explains the entire `StorageAccess` and write-straight-to-physical edge case mechanism.

File: vault/seal_test.go [hunk 14]
```diff
@@ -7,28 +7,31 @@ import (
...
-// of an API.  In other words if your change break this test, it might be more
+// of an API. In other words if your change break this test, it might be more
 // the test's fault than your changes.
 func TestDefaultSeal_Config(t *testing.T) {
 	bc := &SealConfig{
 		SecretShares:    4,
 		SecretThreshold: 2,
 	}
 	core, _, _ := TestCoreUnsealed(t)
+	ctx := namespace.ContextWithNamespace(context.Background(), namespace.RootNamespace)
```
Comment (satoqz) [inline vault/seal_test.go]:
```suggestion
	ctx := namespace.RootContext(nil)
```

File: vault/seal_test.go [hunk 15]
```diff
@@ -39,11 +42,45 @@ func TestDefaultSeal_Config(t *testing.T) {
...
+		SecretShares:    10,
+		SecretThreshold: 5,
+	}
+
+	nsTest := &namespace.Namespace{Path: "test/"}
+	TestCoreCreateNamespaces(t, core, nsTest)
+
+	defSeal = NewDefaultSeal(nil)
+	defSeal.SetCore(core)
+	err = defSeal.SetBarrierConfig(ctx, nsSealConfig, nsTest)
```
Comment (satoqz) [inline vault/seal_test.go]: Only realized now that the type is `SealConfig` but all the functions and fields are "barrier config". If anything this PR is the opportunity to change that (to always be seal config)

File: vault/seal.go [hunk 16]
```diff
@@ -54,14 +55,15 @@ const (
...
 	StoredKeysSupported() seal.StoredKeysSupport // SealAccess
 	SealWrapable() bool
 	SetStoredKeys(context.Context, [][]byte) error
 	GetStoredKeys(context.Context) ([][]byte, error)
-	BarrierType() wrapping.WrapperType                  // SealAccess
-	BarrierConfig(context.Context) (*SealConfig, error) // SealAccess
-	SetBarrierConfig(context.Context, *SealConfig) error
+	BarrierType() wrapping.WrapperType                                        // SealAccess
+	BarrierConfig(context.Context, *namespace.Namespace) (*SealConfig, error) // SealAccess
+	SetBarrierConfig(context.Context, *SealConfig, *namespace.Namespace) error
```
Comment (satoqz) [inline vault/seal.go]: Can we not get the namespace from context? This causes sooo many changed lines just to pass the root namespace in almost all cases.

**Review by satoqz — COMMENTED** [review]
Nothing serious, you could fix all of these later. Looks good generally.

File: command/namespace_create.go (hunk 6 above)
Comment (cipherboy) [reply command/namespace_create.go]: There has to be a better way of parsing this. Maybe this should be added to the API client somehow?

File: vault/init.go [hunk 17]
```diff
@@ -250,7 +250,8 @@ func (c *Core) Initialize(ctx context.Context, initParams *InitParams) (*InitRes
 		return nil, fmt.Errorf("error initializing seal: %w", err)
 	}
 
-	barrierKey, barrierKeyShares, err := c.generateShares(barrierConfig)
+	// TODO: we are here
```
Comment (cipherboy) [inline vault/init.go]: Maybe the TODO should clarify that these shares are only used in tests?

I also wonder if we can drop these tests, given that I think they were pre-Vault 1.0. We don't necessarily support earlier Vault versions and assume you come through v1.14 for migrations, but maybe they were only ever read compatible and not migrated.

File: vault/logical_system_namespaces.go (hunk 8 above)
Comment (cipherboy) [reply vault/logical_system_namespaces.go]: Yeah, I wonder what the most optimal data type would be here. I think the `map name->[]shares` might be best?

Comment (cipherboy) [reply vault/logical_system_namespaces.go]: Or maybe namespaces should only ever be created with a single share and more added later? :thinking:

File: vault/seal.go (hunk 16 above)
Comment (cipherboy) [reply vault/seal.go]: I like this idea FWIW. Worst case, you'll have to find the actual sealed parent namespace but that's about it.

File: vault/seal.go [hunk 18]
```diff
@@ -76,9 +78,10 @@ type Seal interface {
 }
 
 type defaultSeal struct {
-	access seal.Access
-	config atomic.Value
-	core   *Core
+	access     seal.Access
+	config     atomic.Value
+	core       *Core
+	metaPrefix string
```
Comment (cipherboy) [inline vault/seal.go:84]: Not sure if I suggested this elsewhere, but not sure I like the whole `metaPrefix` thing again here, but that's fine.

**Review by cipherboy — COMMENTED** [review]
Some thoughts on the proof of concept, but looking fairly good!

File: command/namespace_create.go (hunk 5 above)
Comment (wslabosz-reply) [reply command/namespace_create.go]: I mean if we want to enable operators to provide a path to config files then yes

Comment (satoqz) [reply command/namespace_create.go]: Can't you just read the relative path if it is relative? I don't see why you need resolve it to an absolute one.

File: vault/seal.go (hunk 16 above)
Comment (wslabosz-reply) [reply vault/seal.go]: that is completely valid, I even adjusted that in my poc multiple seals storage adjustment

File: vault/seal_test.go (hunk 15 above)
Comment (wslabosz-reply) [reply vault/seal_test.go]: `SealConfig` might be a "barrier config" or a "recovery config", so no, this PR is unfortunately not a opportunity to remove that confusion

File: vault/init.go (hunk 17 above)
Comment (wslabosz-reply) [reply vault/init.go]: > Maybe the TODO should clarify that these shares are only used in tests?

this TODO is a leftover from a debug session

> I also wonder if we can drop these tests, given that I think they were pre-Vault 1.0. We don't necessarily support earlier Vault versions and assume you come through v1.14 for migrations, but maybe they were only ever read compatible and not migrated.

valid point, not the scope of this PR per say, but definitely worth a look

File: vault/seal_test.go (hunk 15 above)
Comment (satoqz) [reply vault/seal_test.go]: The point I'm making is that it never configures the barrier, but the sealing mechanism. A "barrier config" does not exist.

Comment (wslabosz-reply) [reply vault/seal_test.go]: okay I get it, so what about
`SetSealConfig` and `SetRecoverySealConfig`
@cipherboy wdyt?

File: command/namespace_create.go (hunk 5 above)
Comment (wslabosz-reply) [reply command/namespace_create.go]: so you are suggesting assuming it is relative, and attempt at reading so, then fallback to absolute?

File: vault/namespace_store.go (hunk 12 above)
Comment (wslabosz-reply) [reply vault/namespace_store.go]: will wait on adjustment after https://github.com/openbao/openbao/pull/1538

File: vault/seal_manager.go (hunk 13 above)
Comment (wslabosz-reply) [reply vault/seal_manager.go:116]: will add something

File: command/namespace_create.go (hunk 5 above)
Comment (satoqz) [reply command/namespace_create.go]: No, just read whatever path is. `os.Open` should understand both types of paths.

File: vault/seal_test.go (hunk 15 above)
Comment (cipherboy) [reply vault/seal_test.go]: Barrier to me refers to keyring specifically, and not to the root keys or seal mechanism that this is.

So I do think I agree the latter set of names are most clear (`SetSealConfig` and `SetRecoverySealConfig`).

File: vault/logical_system_namespaces.go (hunk 8 above)
Comment (wslabosz-reply) [reply vault/logical_system_namespaces.go]: > Or maybe namespaces should only ever be created with a single share and more added later?

would need to think more about it, could we add that to the RFC comments?

File: vault/seal_test.go (hunk 15 above)
Comment (wslabosz-reply) [reply vault/seal_test.go]: I've simplified them even a bit more (removing the `Seal` as it is in the name of the interface and ended up with:
`Config`, `SetConfig`, `SetCachedConfig` (old barrierConfig) and
`RecoveryConfig`, `SetRecoveryConfig`, `SetCachedRecoveryConfig`

File: vault/seal.go (hunk 16 above)
Comment (wslabosz-reply) [reply vault/seal.go]: this caused a ripple effect in tests and some other places, requiring a manual addition of root namespace to the context

File: vault/barrier_aes_gcm.go (hunk 2 above)
Comment (wslabosz-reply) [reply vault/barrier_aes_gcm.go]: @cipherboy
I think it's doable to make it a `physical.View` instead, but I don't think there's a way to leave it with just that, as we need to be able to converse it into `Transactional` barrier, and that requires `physical.Backend`, so we still will be left with:

```
		backend:                  storage,
		view:                     physical.NewView(storage, metaPrefix),
```

> since a physical.BarrierView will be implemented by physical.Barrier

there are no `physical.BarrierView` nor `physical.Barrier` (did you mean `View` and `Backend`?)

Comment (cipherboy) [reply vault/barrier_aes_gcm.go]: Types are soft things that kinda sorta exist in my head but it isn't a type-checking compiler. :laughing:

---

Ahh, right. And we have no way to do a `TransactionalView`. Though, I suppose we could add it...

File: vault/logical_system_namespaces.go (hunk 8 above)
Comment (cipherboy) [reply vault/logical_system_namespaces.go]: @wslabosz-reply Yeah, feel free :-)

File: vault/namespace_store.go (hunk 12 above)
Comment (wslabosz-reply) [reply vault/namespace_store.go]: adjusted

File: vault/barrier_aes_gcm.go (hunk 2 above)
Comment (wslabosz-reply) [reply vault/barrier_aes_gcm.go]: @cipherboy 
should I add it in the scope of this PoC?

Comment (cipherboy) [reply vault/barrier_aes_gcm.go]: Eh, you can add it as a separate PR if you want or we can drop the idea, either way. I suspect there are other places where this'd be useful, for example in @suprjinx's expiration manager transactional view handling, which I've not had time to give feedback on. :(

Comment (wslabosz-reply) [reply vault/barrier_aes_gcm.go]: added it in this PR, not sure if it's correct implementation (I've mirrored the `StorageView` approach)

Comment (phyrog) [reply vault/barrier_aes_gcm.go]: As mentioned in a previous version of this PR, this `metaPrefix` does not construct a view, it just points the barrier to where its config is stored. The barrier will still be able to write to all of the storage (child namespace unseal configs are encrypted with the (next higher sealable) parent namespace barrier).

Comment (phyrog) [reply vault/barrier_aes_gcm.go]: Not sure if I understood correctly, but you can't replace `backend` with a scoped view. The barriers need to be able to write outside of their namespaces path. A namespace's unseal config needs to be encrypted by the next higher sealable namespace barrier, so that barrier needs to be able to write into child namespace data.

What you possibly could do is replace the `metaPrefix` field with a view of just the storage paths that contain the barrier config, while still using the complete `backend` for writing.

File: vault/logical_system_namespaces.go [hunk 19]
```diff
@@ -238,15 +291,21 @@ func (b *SystemBackend) namespacePaths() []*framework.Path {
...
 		"uuid":            ns.UUID,
 		"path":            ns.Path,
 		"id":              ns.ID,
 		"tainted":         ns.Tainted,
 		"locked":          ns.Locked,
 		"custom_metadata": ns.CustomMetadata,
 	}
+	if len(keySharesMap) > 0 {
+		ret["key_shares"] = make(map[string][]string, len(keySharesMap))
+		maps.Copy(ret["key_shares"].(map[string][]string), keySharesMap)
```
Comment (satoqz) [inline vault/logical_system_namespaces.go]: I don't think you need to copy here? `maps.Copy` also won't deepcopy the arrays.

File: sdk/physical/physical_view.go [hunk 20]
```diff
@@ -11,42 +11,80 @@ import (
...
+}
+
+type ViewTransaction interface {
+	Transaction
+	ViewCore
+}
+
+var (
+	_ Backend = &view{}
+	_ View    = &view{}
```
Comment (satoqz) [inline sdk/physical/physical_view.go]: `View` implies `Backend`

File: sdk/physical/physical_view.go [hunk 21]
```diff
@@ -11,42 +11,80 @@ import (
...
+	}
+
+	if _, ok := backend.(Transaction); ok {
+		return &viewTransaction{
+			*v,
+		}
+	}
+
+	return v
 }
```
Comment (satoqz) [inline sdk/physical/physical_view.go]:
```suggestion
func NewView(backend Backend, prefix string) View {
	v := view{
		backend: backend,
		prefix:  prefix,
	}

	if _, ok := backend.(TransactionalBackend); ok {
		return &transactionalView{v}
	}

	if _, ok := backend.(Transaction); ok {
		return &viewTransaction{v}
	}

	return &v
}
```

File: vault/logical_system_namespaces.go (hunk 19 above)
Comment (wslabosz-reply) [reply vault/logical_system_namespaces.go]: right

**Review by cipherboy — APPROVED** [review]
Thank you @wslabosz-reply for your patience and the refactor :-) I think this looks fairly good for the first steps! Really like the use of view and the suggestions from @satoqz, I think it is cleaner.

**Review by satoqz — APPROVED** [review]

File: vault/barrier_aes_gcm.go (hunk 2 above)
Comment (wslabosz-reply) [reply vault/barrier_aes_gcm.go]: @phyrog 
you were right from the beginning, I couldn't make it nicely with the scoped view of the storage, so I've reverted the change.
